src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java

```diff
@@ -0,0 +1,329 @@
+            DataverseXoaiItem xoaiItem = null; 
+            for (OAIRecord oaiRecord : oaiRecords) {
+                if (xoaiItem == null) {
+                    xoaiItem = new DataverseXoaiItem(oaiRecord); 
+                    
+                    // If this is a "deleted" OAI oaiRecord - i.e., if someone
+                    // has called GetRecord on a deleted oaiRecord (??), our 
+                    // job here is done. If it's a live oaiRecord, let's try to 
+                    // look up the dataset and open the pre-generated metadata 
+                    // stream. 
+                    
+                    if (!oaiRecord.isRemoved()) {
+                        Dataset dataset = datasetService.findByGlobalId(oaiRecord.getGlobalId());
+                        if (dataset == null) {
+                            // This should not happen - but if there are no longer datasets 
+                            // associated with this persistent identifier, we should simply 
+                            // bail out. 
+                            // TODO: Consider an alternative - instead of throwing 
+                            // an IdDoesNotExist exception, mark the oaiRecord as 
+                            // "deleted" and serve it to the client (?). For all practical
+                            // purposes, this is what this oaiRecord represents - it's 
+                            // still in the database as part of an OAI set; but the 
+                            // corresponding dataset no longer exists, because it 
+                            // must have been deleted. 
+                            // i.e.
+                            // xoaiItem.getOaiRecord().setRemoved(true);
+                            break;
+                        }
+                        
+                        Metadata metadata;
+                        
+                        try {
+                            metadata = getDatasetMetadata(dataset, metadataFormat.getPrefix());
+                        } catch (ExportException | IOException ex) {
+                            // Again, this is not supposed to happen in normal operations; 
+                            // since by design only the datasets for which the metadata
+                            // records have been pre-generated ("exported") should be 
+                            // served as "OAI Record". But, things happen. If for one
+                            // reason or another that cached metadata file is no longer there, 
+                            // we are not going to serve this oaiRecord. 
+                            // TODO: see the comment above; and consider
+                            // xoaiItem.getOaiRecord().setRemoved(true);
+                            // instead. 
+                            break;
+                        }
+                        xoaiItem.withDataset(dataset).withMetadata(metadata);
```

Comment:
**suggestion:** This very much belongs together - shouldn't all this be inside the `try` block? (Which also could throw directly, as mentioned before/above)

Reply:
I will move the code duplicated between this method and getItems(), and this will be addressed in the process.